### PR TITLE
boards: risc-v: add BeagleV Starlight JH7100 board support 

### PR DESCRIPTION
**Pull request by rajnesh-kanwal**
Adding support for beagleV Starlight board based on Starfive JH7100
SoC. It's a base support, no drivers other than UART has been tested.

Depends on https://github.com/starfive-tech/beagle_ddrinit/pull/8

### CODE REVIEW

File: boards/riscv/beagleV/doc/index.rst [hunk 1]
```diff
@@ -0,0 +1,71 @@
+.. _beagelV:
```
Comment (cfriedt) [inline boards/riscv/beagleV/doc/index.rst:1]: Typo

File: boards/riscv/beagleV/doc/index.rst [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+*************************
+
+Building
+========
+
+Applications for the ``beagleV`` board configuration can be built as usual
+(see :ref:`build_an_application`) using the corresponding board name:
+
+.. zephyr-app-commands::
+   :board: beagleV
```
Comment (cfriedt) [inline boards/riscv/beagleV/doc/index.rst:22]: In general, I don't see many boards (if any) using camel case. I think most of them are all lower case.

Perhaps someone else can chime in to see if lowercase board names are a policy or convention or just coincidence.

File: soc/riscv/riscv-privilege/starfive_jh71xx/Kconfig.soc [hunk 3]
```diff
@@ -0,0 +1,11 @@
+# SPDX-License-Identifier: Apache-2.0
+
+choice
+	prompt "RISC-V Starfive JH7100 SoC"
```
Comment (katsuster) [inline soc/riscv/riscv-privilege/starfive_jh71xx/Kconfig.soc:4]: RISC-V is obvious so "Starfive JH7100 SoC" is better.

File: soc/riscv/riscv-privilege/starfive_jh71xx/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,3 @@
+# SPDX-License-Identifier: Apache-2.0
```
Comment (robertlipe) [inline soc/riscv/riscv-privilege/starfive_jh71xx/CMakeLists.txt:1]: Name your directories for a day when a (potentially annoyingly incompatible) successor for JH7100 exists. Yes, that means busywork in the future IF the chip is fully compatible.

**Review by robertlipe — COMMENTED** [review]
Minor nits for style on a drive-by review. Feel free to ignore me.
Congrats for getting Zephyr up.  At one time, I considered doing this. :-)

File: soc/riscv/riscv-privilege/starfive_jh71xx/soc.h [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+ */
+
+#ifndef __RISCV_VIRT_SOC_H_
+#define __RISCV_VIRT_SOC_H_
+
+#include <soc_common.h>
+#include <devicetree.h>
+
+#define RISCV_MTIME_BASE 0x0200BFF8
+#define RISCV_MTIMECMP_BASE 0x02004000
```
[the remark this answers is not in the document]
Comment (rajnesh-kanwal) [reply soc/riscv/riscv-privilege/starfive_jh71xx/soc.h:14]: Yes, it can be derived from DTS as Linux is doing. but the problem is it's being used by generic timer driver `drivers/timer/riscv_machine_timer.c` and this macro is being defined by other ports as well. Obviously, as long as I keep the name the same, it won't break anything but for the sake of consistency between different ports I was planning to do this improvement in a separate PR.

File: soc/riscv/riscv-privilege/starfive_jh71xx/CMakeLists.txt (hunk 4 above)
Comment (rajnesh-kanwal) [reply soc/riscv/riscv-privilege/starfive_jh71xx/CMakeLists.txt:1]: I am not sure about this, to be honest. Compatibility looks like an issue given how things are going on.
But I guess we can keep the directory name the same, as even there might be compatibility issues, these SoC belong to the same series and I think it can be tackled with different configs and conditional code and stuff. 
I am not really sure creating two directories like `starfive_jh7100` and `starfive_jh7110` would be a nice idea.

File: boards/riscv/beagleV/doc/index.rst [hunk 6]
```diff
@@ -0,0 +1,71 @@
+.. _beagelV:
+
+BeagleV
+#######
+
+Overview
+********
+
+The BeagleV is an 64-bit open-source RISC-V development board with
```
[the remark this answers is not in the document]
Comment (rajnesh-kanwal) [reply boards/riscv/beagleV/doc/index.rst:9]: Thanks for pointing it out. I will fix this in the next update.

File: boards/riscv/beagleV/doc/index.rst (hunk 2 above)
Comment (rajnesh-kanwal) [reply boards/riscv/beagleV/doc/index.rst:22]: Yes beagle_v sounds like a nice compromise if we can't have camel case names. 
One more thing, I guess it should be bealge_v_evb and for the production board, it should be simple beagle_v. 
it would be great if someone can confirm the camel case thing.

File: boards/riscv/beagleV/doc/index.rst (hunk 6 above)
[the remark this answers is not in the document]
Comment (cfriedt) [reply boards/riscv/beagleV/doc/index.rst:9]: It might be good to proactively change all of the board names to `beagle_v_starlight` since there is an indication more are on the way

File: soc/riscv/riscv-privilege/starfive_jh71xx/CMakeLists.txt (hunk 4 above)
Comment (cfriedt) [reply soc/riscv/riscv-privilege/starfive_jh71xx/CMakeLists.txt:1]: I think `starfive_jh71xx` should be ok as well - ideally any minor inconsistencies should be handled either by devicetree compatibility strings / macros or (in the worst case) some runtime checks.

**Comment by pdp7** [comment]
> Adding support for beagleV board based on Starfive VIC7100
> SoC. It's a base support, no drivers other than UART has been
> tested.
> 
> Depends on [starfive-tech/beagle_ddrinit#8](https://github.com/starfive-tech/beagle_ddrinit/pull/8)

a note on naming: BeagleV is a category term that refers to all RISC-V boards from BeagleBoard.org, BeagleV Starlight is the name of BeagleV boards with StarFive SoCs.  There are two version: BeagleV Starlight JH7100 (the beta prototype) and BeagleV Starlight JH7110 (future production version)

https://github.com/beagleboard/beaglev-starlight#guidance-on-naming-conventions

File: boards/riscv/beagleV/doc/index.rst (hunk 1 above)
Comment (rajnesh-kanwal) [reply boards/riscv/beagleV/doc/index.rst]: Fixed in e7ce67a4ca44281ea4d39cb4a7449b113738b8f3

File: boards/riscv/beagleV/doc/index.rst (hunk 2 above)
Comment (rajnesh-kanwal) [reply boards/riscv/beagleV/doc/index.rst]: Changed the name to beagle_v_starlight.

File: soc/riscv/riscv-privilege/starfive_jh71xx/Kconfig.soc (hunk 3 above)
Comment (rajnesh-kanwal) [reply soc/riscv/riscv-privilege/starfive_jh71xx/Kconfig.soc]: Fixed.
Thanks.

File: boards/riscv/beagleV/doc/index.rst (hunk 6 above)
[the remark this answers is not in the document]
Comment (rajnesh-kanwal) [reply boards/riscv/beagleV/doc/index.rst]: Changed the board name, options and filenames to beagle_v_starlight and board mentions in docs to beagleV starlight JH7100.

File: boards/riscv/beagle_v_starlight/Kconfig.board [hunk 7]
```diff
@@ -0,0 +1,6 @@
+# SPDX-License-Identifier: Apache-2.0
+
+config BOARD_BEAGLE_V_STARLIGHT
```
Comment (robertlipe) [inline boards/riscv/beagle_v_starlight/Kconfig.board:3]: I'm not sure how much of this naming dirty laundry to make visible to end users, but we'll have at least two "BeagleV Starlight" boards: one with 7100 and one with 7110. Does this naming scheme allow that?

(I'm struggling with this in my own code, too....)

File: soc/riscv/riscv-privilege/starfive_jh71xx/CMakeLists.txt (hunk 4 above)
Comment (robertlipe) [reply soc/riscv/riscv-privilege/starfive_jh71xx/CMakeLists.txt:1]: I'm hip with this. Thanx!

**Review by robertlipe — COMMENTED** [review]
My happiness is of no consequence, but I like this much better. Thanx for humoring me.

File: boards/riscv/beagle_v_starlight/Kconfig.board (hunk 7 above)
Comment (rajnesh-kanwal) [reply boards/riscv/beagle_v_starlight/Kconfig.board:3]: beagle_v_starlight_beta sounds lengthy, to be honest. I am not sure about this.

I am kind of under the impression that this board is for development purpose only and there won't be any more boards based on JH7100. So I don't think it makes sense to continue supporting this beta SoC. Instead, the work done on it can be used as a base and when the production SoC comes in, it's ahead of the initial phase of development.

So in short what I am suggesting here is, we can keep the name beagle_v_starlight. when the production board comes in, this port can be migrated to support the production version and we can say goodbye to the beta one.
Just some raw thoughts. @pdp7, can you comment on this?

Comment (pdp7) [reply boards/riscv/beagle_v_starlight/Kconfig.board:3]: I think the config option should be BEAGLEV_STARLIGHT_JH7100.  I would not use the terms beta or production.  Instead, it is best to use: BeagleV Starlight JH7100 and BeagleV Starlight JH7110.  

When adding support for JH7110 in the future, it may be possible to have an a wildcard config BEAGLEV_STARLIGHT_JH71XX to enable common support and then BEAGLEV_STARLIGHT_JH7100 and BEAGLEV_STARLIGHT_JH7110 depend on  BEAGLEV_STARLIGHT_JH71XX being enabled

**Review by katsuster — APPROVED** [review]

File: boards/riscv/beagleV/doc/index.rst (hunk 6 above)
[the remark this answers is not in the document]
Comment (cfriedt) [reply boards/riscv/beagleV/doc/index.rst]: @pdp7 - is `beagle_v_starlight` ok with you?

[the remark this answers is not in the document]
Comment (cfriedt) [reply boards/riscv/beagleV/doc/index.rst]: Ugh - sorry to ask for more churn, but 

https://github.com/beagleboard/beaglev-starlight#guidance-on-naming-conventions

mentions that boards should be differentiated

`beaglev_starlight_jh7100` vs `beaglev_starlight_jh7110`

**Comment by cfriedt** [comment]
> a note on naming: BeagleV is a category term that refers to all RISC-V boards from BeagleBoard.org, BeagleV Starlight is the name of BeagleV boards with StarFive SoCs. There are two version: BeagleV Starlight JH7100 (the beta prototype) and BeagleV Starlight JH7110 (future production version)
> 
> https://github.com/beagleboard/beaglev-starlight#guidance-on-naming-conventions

Thanks for specifying naming conventions, @pdp7

File: boards/riscv/beagle_v_starlight/Kconfig.board (hunk 7 above)
Comment (cfriedt) [reply boards/riscv/beagle_v_starlight/Kconfig.board:3]: @rajnesh-kanwal - @pdp7 is the authoritative source ^^ ;-)

uppercase for Kconfig options, and then directly to lowercase for the board name, paths, filenames, etc.

File: boards/riscv/beagle_v_starlight/Kconfig.defconfig [hunk 8]
```diff
@@ -0,0 +1,5 @@
+# SPDX-License-Identifier: Apache-2.0
+
+config BOARD
+	default "beagle_v_starlight"
```
Comment (cfriedt) [inline boards/riscv/beagle_v_starlight/Kconfig.defconfig:4]: as per @pdp7's recommendation, I would change to `beaglev_starlight_jh7100` and rename the parent directory accordingly.

File: boards/riscv/beagle_v_starlight/beagle_v_starlight.dts [hunk 9]
```diff
@@ -0,0 +1,21 @@
+/* SPDX-License-Identifier: Apache-2.0 */
+
+/dts-v1/;
+
+#include "starfive/starfive_jh7100_beagle_v.dtsi"
```
Comment (cfriedt) [inline boards/riscv/beagle_v_starlight/beagle_v_starlight.dts:5]: as per @pdp7's recommendation, I would change to `beaglev_starlight_jh7100`

Also, please change to the recommended compat strings

https://github.com/beagleboard/beaglev-starlight#guidance-on-naming-conventions

File: boards/riscv/beagle_v_starlight/beagle_v_starlight.yaml [hunk 10]
```diff
@@ -0,0 +1,9 @@
+identifier: beagle_v_starlight
```
Comment (cfriedt) [inline boards/riscv/beagle_v_starlight/beagle_v_starlight.yaml:1]: as per @pdp7's recommendation, I would change to `beaglev_starlight_jh7100`

File: boards/riscv/beagle_v_starlight/beagle_v_starlight_defconfig [hunk 11]
```diff
@@ -0,0 +1,23 @@
+# SPDX-License-Identifier: Apache-2.0
```
Comment (cfriedt) [inline boards/riscv/beagle_v_starlight/beagle_v_starlight_defconfig:1]: as per @pdp7's recommendation, I would change to `beaglev_starlight_jh7100`

File: boards/riscv/beagle_v_starlight/board.cmake [hunk 12]
```diff
@@ -0,0 +1 @@
+# SPDX-License-Identifier: Apache-2.0
```
Comment (cfriedt) [inline boards/riscv/beagle_v_starlight/board.cmake:1]: as per @pdp7's recommendation, I would change to `beaglev_starlight_jh7100`

File: boards/riscv/beagle_v_starlight/doc/index.rst [hunk 13]
```diff
@@ -0,0 +1,75 @@
+.. _beagle_v_starlight:
```
Comment (cfriedt) [inline boards/riscv/beagle_v_starlight/doc/index.rst:1]: as per @pdp7's recommendation, I would change to `beaglev_starlight_jh7100`

File: dts/riscv/starfive/starfive_jh7100_beagle_v.dtsi [hunk 14]
```diff
@@ -0,0 +1,190 @@
+#include "starfive_jh7100_clk.dtsi"
```
Comment (cfriedt) [inline dts/riscv/starfive/starfive_jh7100_beagle_v.dtsi:1]: It's possible that this might be better in the board directory rather than `dts/`

File: boards/riscv/beagle_v_starlight/Kconfig.defconfig (hunk 8 above)
Comment (rajnesh-kanwal) [reply boards/riscv/beagle_v_starlight/Kconfig.defconfig]: Fixed.

File: boards/riscv/beagle_v_starlight/beagle_v_starlight.dts (hunk 9 above)
Comment (rajnesh-kanwal) [reply boards/riscv/beagle_v_starlight/beagle_v_starlight.dts]: Fixed.

File: boards/riscv/beagle_v_starlight/beagle_v_starlight.yaml (hunk 10 above)
Comment (rajnesh-kanwal) [reply boards/riscv/beagle_v_starlight/beagle_v_starlight.yaml]: Fixed.

File: boards/riscv/beagle_v_starlight/beagle_v_starlight_defconfig (hunk 11 above)
Comment (rajnesh-kanwal) [reply boards/riscv/beagle_v_starlight/beagle_v_starlight_defconfig]: Fixed.

File: boards/riscv/beagle_v_starlight/board.cmake (hunk 12 above)
Comment (rajnesh-kanwal) [reply boards/riscv/beagle_v_starlight/board.cmake]: Fixed.

File: boards/riscv/beagle_v_starlight/doc/index.rst (hunk 13 above)
Comment (rajnesh-kanwal) [reply boards/riscv/beagle_v_starlight/doc/index.rst]: Fixed.

File: dts/riscv/starfive/starfive_jh7100_beagle_v.dtsi (hunk 14 above)
Comment (rajnesh-kanwal) [reply dts/riscv/starfive/starfive_jh7100_beagle_v.dtsi:1]: I am not sure about this one tbh. I think this dts belongs to soc rather than the board. I can rename the dts file if that makes sense. Something like starfive_jh7100.dtsi. This one is taken from [here](https://github.com/starfive-tech/u-boot/tree/Fedora_VIC_7100_2021.04/arch/riscv/dts) and so the name follows.

File: boards/riscv/beagleV/doc/index.rst (hunk 6 above)
[the remark this answers is not in the document]
Comment (rajnesh-kanwal) [reply boards/riscv/beagleV/doc/index.rst]: Fixed.

File: soc/riscv/riscv-privilege/starfive_jh71xx/soc.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (cfriedt) [reply soc/riscv/riscv-privilege/starfive_jh71xx/soc.h:14]: I think this is ok. Other SoC's are doing the same thing atm in `soc.h`. It would be good if it was brought out to DT later, as long as complementary changes can be made for other SoC's as well.

File: soc/riscv/riscv-privilege/starfive_jh71xx/Kconfig.soc (hunk 3 above)
Comment (cfriedt) [reply soc/riscv/riscv-privilege/starfive_jh71xx/Kconfig.soc]: Resolving as @katsuster has already OK'ed

**Review by cfriedt — APPROVED** [review]

**Comment by nashif** [comment]
Can this board be purchased somewhere?

**Comment by robertlipe** [comment]
BeagleV Starlight is currently in the hands of developers working on
projects like this. (Minutes ago, I PR'ed a similar commit to another
Apache operating system for this hardware.)
Orders for the actual board have not yet begun.
https://beagleboard.org/beaglev will have more information when the time is
right.
